src/Api/NavigationUtils.ts

```diff
@@ -150,9 +177,216 @@ export const getResolveConditionsResult = (
+  if (tree.routeName === routeName) {
+    return tree
+  }
+  if ('screens' in tree) {
+    for (const screen of tree.screens) {
+      let foundScreen
+      if (screen.routeName === routeName) {
+        foundScreen = screen
+      }
+      if ('screens' in screen) {
+        foundScreen = findStaticScreenTree(screen, routeName)
+      }
+      if (foundScreen != null) {
+        return foundScreen
+      }
+    }
+  }
+}
+
+export const getRouteNameForStateKey = (
+  state: NavigationState,
+  stateKey: string,
+  isRoot = true,
+): string | undefined => {
+  if (isRoot && state.key === stateKey) {
+    return ROOT_NAVIGATOR_ID
+  }
+  const { routes } = state
+  for (const route of routes) {
+    if (route.state?.key === stateKey) {
+      return route.name
+    }
+    if (route.state?.routes != null) {
+      const routeName = getRouteNameForStateKey(route.state as NavigationState, stateKey, false)
+      if (routeName != null) {
+        return routeName
+      }
+    }
+  }
+}
+
+export const findStaticNavigatorForStateKey = (
+  tree: StaticScreenTreeWithDepth,
+  state: NavigationState,
+  stateKey: string,
+): StaticScreenTreeNavigatorWithDepth | undefined => {
+  const routeName = getRouteNameForStateKey(state, stateKey)
+  if (routeName == null) {
+    return undefined
+  }
+  return findStaticScreenTree(tree, routeName) as StaticScreenTreeNavigatorWithDepth
+}
+
+const createParams = (path: string[]): Params | undefined => {
+  if (path.length === 0) {
+    return undefined
+  }
+  const [routeName, ...restPath] = path
+  if (restPath.length === 0) {
+    return {
+      screen: routeName,
+    }
+  }
+  return {
+    screen: routeName,
+    params: createParams(restPath),
+  }
+}
+
+const createNavigateActionForPath = (path: string[], originalAction: NavigationAction): NavigationAction => {
+  const pathFromAction = getRoutePathFromAction(originalAction)
+  const [routeName, ...restPath] = path
+  const nextPath = pathFromAction != null ? [...restPath, ...pathFromAction] : restPath
+  return {
+    type: 'NAVIGATE',
+    payload: {
+      name: routeName,
+      params: createParams(nextPath),
+    },
+  }
+}
+
+const arePathsStartingEqually = (leftPath: string[], rightPath: string[]): boolean => {
+  const shorter = leftPath.length < rightPath.length ? leftPath : rightPath
+  for (let i = 0; i < shorter.length; i++) {
+    if (leftPath[i] !== rightPath[i]) {
+      return false
+    }
+  }
+  return true
+}
+
+const getCommonStaticNavigatorWithPaths = ({
+  currentStaticTreeScreen,
+  finalStaticTreeScreen,
+}: {
+  currentStaticTreeScreen?: StaticScreenTreeWithDepth,
+  finalStaticTreeScreen?: StaticScreenTreeWithDepth,
+}): {
+  commonStaticNavigator: StaticScreenTreeNavigatorWithDepth | undefined,
+  targetPathFromCommonNavigator: string[],
+  sourcePathFromCommonNavigator: string[],
+} => {
+  let leftDepth = currentStaticTreeScreen?.depth ?? 0
+  let rightDepth = finalStaticTreeScreen?.depth ?? 0
```

Comment:
do we need to continue with logic below if one them is not present, we have lot of optional chaining bellow.